flexmeasures/data/services/data_sources.py

```diff
@@ -13,15 +16,22 @@ def get_or_create_source(
     source_type: str | None = None,
     model: str | None = None,
     version: str | None = None,
+    attributes: dict | None = None,
     flush: bool = True,
 ) -> DataSource:
+    attributes_hash = None
     if is_user(source):
         source_type = "user"
     query = DataSource.query.filter(DataSource.type == source_type)
     if model is not None:
         query = query.filter(DataSource.model == model)
     if version is not None:
         query = query.filter(DataSource.version == version)
+    if attributes is not None:
+        attributes_hash = hashlib.sha256(
```

Comment:
Maybe this could become a static function `DataSource.hash_attributes` for code re-use